feat: create newsapi toyproject

Hungeun2 opened the pull request:
## 📚 개요
newsapi 프로젝트 Vanilla JS.ver

## 📚 작업사항 (기능추가/수정/삭제 등)
- [프로젝트 생성] 

## 📚 중요사항

## 📚 궁금한 점

## 📚 어려웠던 점

## 📚 비고

Hungeun2 committed:
feat: create newsapi toyproject

Hungeun2 committed:
fix change directory

Hungeun2 committed:
docs: edit gitignore

Hungeun2 committed:
fix: index.html 오타수정

danbileee commented on week6/hun-geun/newsapi/main.js:
```diff
@@ -0,0 +1,92 @@
+    const keyWord = inputWord.value.trimStart()
+    keyWord ? getNewsData(keyWord) : alert("검색어를 입력하세요.")
+    inputWord.value = '';
+    const resultKey = document.querySelector('.result-key')
+    resultKey.innerHTML = `Result about '${keyWord}'`;
```
여기 함수명이 `search` 여서 해당 구문은 `render` 쪽으로 이동한다면 좀 더 역할 구분이 될 것 같습니다!

Seok93 replied:
멘토님이 말씀해주신 SOLID 원칙의 S에 해당하는 내용입니다.
S는 단일 책임의 원칙(SRP, Single Responsibility Principle)으로 하나의 모듈은 한 가지 책임을 가져야 한다는 것을 의미합니다.  검색을 해주는 역할과 화면에 그려주는 렌더의 역활을 더 확실하게 분리해주면 유지보수에 도움이 됩니다,

[객체지향 프로그래밍의 5가지 설계 원칙, SOLID](https://mangkyu.tistory.com/194) 참고 자료

Seok93 replied:
[코드가 깔끔해지는 3가지 원칙](https://www.youtube.com/watch?v=jafa3cqoAVM&t=4s) SOLID원칙이랑 더불어서 알고 있으면 좋은 개념들 영상 추가합니다.

danbileee commented on week6/hun-geun/newsapi/main.js:
```diff
@@ -0,0 +1,92 @@
+        throw new Error("Can't Search News");
+    } 
+    const newsData = await ori.json();
+    if (newsData.articles.length === 0){
+        alert(`${word}에 대한 검색결과가 없습니다.`)
+    } else {
+        dataProcess(newsData.articles);
+    }
```
이쪽도 함수명과 관련하여, `getNewsData` 이면 응답값 리턴까지만 수행하면 될 것 같아요!
응답값에 따른 화면 처리를 담당하는 역할은 적절한 다른 함수나 이벤트 핸들러 내에서 수행해볼 수 있지 않을까 싶습니다🙂

danbileee commented on week6/hun-geun/newsapi/main.js:
```diff
@@ -0,0 +1,92 @@
+
+/* 항목별 분류 */
+function dataProcess(data) {
+    let authorList = [];
+    let titleList = [];
+    let urlList = [];
+    let urlToImageList = [];
+    let publishedAtList = [];
+
+    data.forEach(el =>{
+        authorList.push(el.author);
+        titleList.push(el.title);
+        urlList.push(el.url);
+        urlToImageList.push(el.urlToImage);
+        publishedAtList.push(el.publishedAt.slice(0,10));
+    });
+    render(authorList, titleList, urlList, urlToImageList, publishedAtList);
```
항목별로 분류해주시는 것도 가능한 선택이나 아무래도 데이터 필드가 늘어난다거나 코드베이스가 방대해질수록 관리는 어려워질 것 같습니다.
API 응답으로 받은 객체 구조는 가능한 한 유지하고, 이 객체의 배열을 순회하여 필요한 처리를 해보시는 방향으로 한번 개선해보시면 어떨까 싶어요!
이와 관련해서 각각의 항목을 배열로 분리하다 보니 이어지는 render 함수의 인자가 5개정도인데 토이프로젝트임을 감안해도 충분히 복잡도가 높다고 느껴지는 인터페이스 같습니다.

Hungeun2 replied:
해당 코드를 작성하면서도 의문이 많이 들었던 부분이었는데 제시해 주신 방법으로 개선해 보겠습니다!!

danbileee commented on week6/hun-geun/newsapi/main.js:
```diff
@@ -0,0 +1,92 @@
+    ulList.innerHTML='';
+    /*Error prevention*/
+    const resultNumber = document.querySelector('.result-number');
+    try {
+        const showResult = document.querySelector('.result');
+        showResult.className = 'result-show';
+    } catch{}
```
try catch 가 필요하지 않을듯한 구문같아요!
혹시 `showResult` 값이 없는 상황에 대한 방어코드가 필요하다면 조건문이나 optional chaining을 활용해보셔도 될 것 같습니다.

danbileee commented on week6/hun-geun/newsapi/main.js:
```diff
@@ -0,0 +1,92 @@
+        showResult.className = 'result-show';
+    } catch{}
+
+    authorList.length <= 20 ? repeat = authorList.length : repeat = 20;
```
재할당이 일어난다기 보다는 조건에 따라 다른 값을 할당하는 구문이어서 `repeat` 을 let 변수가 아닌 const로 선언 후 삼항연산자 사용하시는 게 어떨까 싶습니다.
```suggestion
    const reapeat = authorList.length <= 20 ? authorList.length : 20;
```

danbileee reviewed (commented):
수고 많으셨습니다!

Hungeun2 committed:
feat: main.css 수정
result-show class 삭제

Hungeun2 committed:
feat: main.js 수정
search함수 수정 -> inputCheck 및 newsHandle로 세분화
dataProcess 함수 삭제
render함수에서 try catch문 삭제, showResult 구문 삭제
repeat 할당 부분 let에서 const로 변경
